File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
    if (!root) return []
    const levels = []
    let minCol = 0
    let maxCol = 0
    let colMap = new Map()

    const queue = [{node: root, row: 0, col: 0}]

    while (queue.length > 0) {
        const { node, row, col } = queue.shift()

        if (!colMap.has(col)) {
            colMap.set(col, [])
        }
        colMap.get(col).push([row, node.val])

        minCol = Math.min(minCol, col)
        maxCol = Math.max(maxCol, col)

        if (node.left) {
            queue.push({ node: node.left, row: row + 1, col: col - 1})
        }
        if (node.right) {
            queue.push({ node: node.right, row: row + 1, col: col + 1})
        }
    }

    const result = []
    for (let c = minCol; c <= maxCol; c++) {
        const pairs = colMap.get(c) || []
        pairs.sort((a, b) => {
            if (a[0] !== b[0]) return a[0] - b[0]
            return a[1] - b[1]
        })
        result.push(pairs.map(pair => pair[1]))
    }

    return result
};

File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        column_map = defaultdict(list)
        queue = deque([(root, 0, 0)])
        min_col = max_col = 0

        while queue:
            node, row, column = queue.popleft()
            min_col = min(min_col, column)
            max_col = max(max_col, column)

            column_map[column].append((row, node.val))

            if node.left:
                queue.append((node.left, row + 1, column - 1))
            if node.right:
                queue.append((node.right, row + 1, column + 1))

        res = []

        for i in range(min_col, max_col + 1):
            column = [val for _, val in sorted(column_map[i], key=lambda x: (x[0], x[1]))]
            res.append(column)

        return res